src/coreclr/jit/jiteh.cpp

```diff
@@ -2506,6 +2512,82 @@ bool Compiler::fgNormalizeEHCase2()
     return modified;
 }
 
+//------------------------------------------------------------------------
+// fgNormalizeEHCase_NativeAot:
+//     For Exception types which require runtime lookup it creates a "fake" single-block
+//     EH filter that performs "catchArg isinst T!!" and in case of success forwards to the
+//     original EH handler.
+//
+// Return Value:
+//    true if basic block layout was changed
+//
+
+bool Compiler::fgNormalizeEHCase_NativeAot()
+{
+    // CI test, let's run the logic for JIT
+    //assert(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+
+    bool modified = false;
+    for (unsigned XTnum = 0; XTnum < compHndBBtabCount; XTnum++)
+    {
+        EHblkDsc* eh = ehGetDsc(XTnum);
+        if ((eh->ebdHandlerType == EH_HANDLER_CATCH) && !eh->HasFilter())
+        {
+            CORINFO_RESOLVED_TOKEN resolvedToken;
+            resolvedToken.tokenContext = impTokenLookupContextHandle;
+            resolvedToken.tokenScope   = info.compScopeHnd;
+            resolvedToken.token        = eh->ebdTyp;
+            resolvedToken.tokenType    = CORINFO_TOKENKIND_Casting;
+            info.compCompHnd->resolveToken(&resolvedToken);
+
+            bool     runtimeLookupNeeded = false;
+            GenTree* runtimeLookup       = impTokenToHandle(&resolvedToken, &runtimeLookupNeeded, false);
+            if (!runtimeLookupNeeded)
+            {
+                // Exception type does not need runtime lookup
+                continue;
+            }
+
+            // Create a new bb for the fake filter
+            BasicBlock* handlerBb = eh->ebdHndBeg;
+            BasicBlock* filterBb  = bbNewBasicBlock(BBJ_EHFILTERRET);
+            filterBb->bbCatchTyp  = BBCT_FILTER;
+            filterBb->bbCodeOffs  = handlerBb->bbCodeOffs; // Technically, we're in the handler
+            filterBb->bbHndIndex  = handlerBb->bbHndIndex;
+            filterBb->bbTryIndex  = handlerBb->bbTryIndex;
+            filterBb->bbJumpDest  = handlerBb;
+            filterBb->bbSetRunRarely();
+            filterBb->bbFlags |= BBF_INTERNAL | BBF_DONT_REMOVE;
+
+            // Now we need to spill CATCH_ARG (it should be the first thing evaluated)
+            GenTree* arg = new (this, GT_CATCH_ARG) GenTree(GT_CATCH_ARG, TYP_REF);
+            arg->gtFlags |= GTF_ORDER_SIDEEFF;
+            unsigned tempNum         = lvaGrabTemp(false DEBUGARG("SpillCatchArg"));
+            lvaTable[tempNum].lvType = TYP_REF;
+            GenTree* argAsg          = gtNewTempAssign(tempNum, arg);
+            arg                      = gtNewLclvNode(tempNum, TYP_REF);
+            filterBb->bbStkTempsIn   = tempNum;
+            fgInsertStmtAtBeg(filterBb, gtNewStmt(argAsg));
+
+            // Create "catchArg is TException" tree
+            GenTree* isInstOfT = gtNewHelperCallNode(CORINFO_HELP_ISINSTANCEOFANY, TYP_REF, runtimeLookup, arg);
```

Comment:
Exception handling uses a simple assignability check that just walks the inheritance chain:
 https://github.com/dotnet/runtime/blob/f19d85b2b6673312c557293c5aadaac11dab84d7/src/coreclr/vm/excep.cpp#L159

It may be interesting to fix it to just use a regular assignability check, but not as part of this change. This change may need a new helper for this.

Reply:
I guess we do not have a test that would hit this problem. This should get a new test and get addressed.

Reply:
@jkotas could you please explain - what's wrong with the  `CORINFO_HELP_ISINSTANCEOFANY`?
this basically does:
```
  IL_0014:  isinst     !!T
  IL_0019:  ldnull
  IL_001a:  cgt.un
  IL_001f:  endfilter
```

Reply:
do you mean that the catch_arg can be a value type?

Reply:
The problem is that `CORINFO_HELP_ISINSTANCEOFANY` does not match what the `ExceptionIsOfRightType` method used by non-generic exception handling does to match types. It is not possible to demonstrate the observable behavior difference in C# since C# requires caught type to inherit from `System.Exception`. The behavior difference can be observed in IL and also possibly in VB and managed C++.

Here is IL repro: 

https://gist.github.com/jkotas/09cef5b3efd4d617ae388f0eda9a06ef

`ilasm repro.il /out:repro.dll`

It prints `PASS PASS` in main, but it is going to print `PASS FAIL` with your change.

Reply:
Thanks! I'll make a new helper then